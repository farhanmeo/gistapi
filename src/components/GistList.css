.gist-list {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    min-height: 80vh; /* Set minimum height to cover the viewport */
  }
  
  /* Style for the message */
  .no-data {
    text-align: center;
  }
  .no-data-image {
    width: 500px; /* Adjust the size as needed */
    margin-bottom: 20px;
  }