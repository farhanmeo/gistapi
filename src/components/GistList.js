import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { fetchGists } from "../actions/gistActions";
import Gist from "./Gist";
import "./GistList.css";

const GistList = ({ searchTerm }) => {
  const dispatch = useDispatch();
  const allGists = useSelector((state) => state.gists.gists);
  const [filteredGists, setFilteredGists] = useState(allGists);

  useEffect(() => {
    const fetchData = async () => {
      try {
        await dispatch(fetchGists()); // Fetch all gists by default
      } catch (error) {
        // Handle the error if needed
        console.error("Error fetching gists:", error);
      }
    };
  
    fetchData();
  }, [dispatch]);
  useEffect(() => {
    if (searchTerm) {
      const filtered = allGists.filter((gist) =>
        gist.owner.login.toLowerCase().includes(searchTerm.toLowerCase())
      );
      setFilteredGists(filtered);
    } else {
      setFilteredGists(allGists); // Show all gists if no search term
    }
  }, [allGists, searchTerm]);
  console.log("filteredGists", filteredGists);
  return (
    <div >
      {filteredGists.length > 0 ? (
        filteredGists.map((gist) => <Gist key={gist.id} gist={gist} />)
      ) : (
        <div className="gist-list no-data">
          <img
            src="/assets/noRecords.png"
            alt="No Data"
            className="no-data-image"
          />
        </div>
      )}
    </div>
  );
};

export default GistList;
