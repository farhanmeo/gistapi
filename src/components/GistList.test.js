import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store"; // Import the redux-mock-store package
import GistList from "./GistList"; // Import the component you want to test

const mockStore = configureStore([]); // Create a mock store

describe("GistList Component UI Rendering", () => {
  it("should render no data message when filteredGists is empty", () => {
    const initialState = {
      gists: {
        gists: [], // Simulate empty gists array
      },
    };
    const store = mockStore(initialState); // Create a mock store with the initialState

    // Render the GistList component wrapped in the Provider with the mock store
    const { getByAltText } = render(
      <Provider store={store}>
        <GistList searchTerm="" />
      </Provider>
    );

    // Assert that the "No Data" image 
    const noDataImage = getByAltText("No Data");
    expect(noDataImage).toBeInTheDocument();
  });

  it('should render Gist components when filteredGists has data', () => {
    const initialState = {
      gists: {
        gists: [
          // Simulate some gists data
          {
            id: '1',
            owner: { login: 'user1' },
          },
          {
            id: '2',
            owner: { login: 'user2' },
          },
        ],
      },
    };
    const store = mockStore(initialState);
  
    const { getAllByTestId } = render(
      <Provider store={store}>
        <GistList searchTerm="" />
      </Provider>
    );
  
    const gistComponents = getAllByTestId('gist-component');
    expect(gistComponents).toHaveLength(2); // Number of gists in the initial state
  });
    

  it('should filter gists based on search term', () => {
    const initialState = {
      gists: {
        gists: [
          {
            id: '1',
            owner: { login: 'user1' },
          },
          {
            id: '2',
            owner: { login: 'user2' },
          },
        ],
      },
    };
    const store = mockStore(initialState);
  
    const { getByText, queryByText } = render(
      <Provider store={store}>
        <GistList searchTerm="user1" />
      </Provider>
    );
  
    const gistUser1 = getByText('user1');
    const gistUser2 = queryByText('user2');
    expect(gistUser1).toBeInTheDocument();
    expect(gistUser2).not.toBeInTheDocument();
  });
  });
