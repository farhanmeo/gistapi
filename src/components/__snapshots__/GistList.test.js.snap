// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GistList Component should match snapshot when filteredGists has data 1`] = `
<DocumentFragment>
  <div>
    <div
      class="gist"
    >
      <div
        class="gist-header"
      >
        <div
          class="user-info"
        >
          <img
            alt="user1"
            class="user-avatar"
          />
          <h2>
            user1
          </h2>
        </div>
        <div
          class="user-details"
        >
          <div
            class="gist-links"
          >
            <a
              rel="noreferrer"
              target="_blank"
            >
              <span
                class="octicon octicon-file"
              />
                Files
            </a>
            <a
              rel="noreferrer"
              target="_blank"
            >
              <span
                class="octicon octicon-repo-forked"
              />
                Forks
            </a>
            <a
              rel="noreferrer"
              target="_blank"
            >
              <span
                class="octicon octicon-comment"
              />
                Comments
            </a>
            <a
              rel="noreferrer"
              target="_blank"
            >
              <span
                class="octicon octicon-star"
              />
                Stars
            </a>
          </div>
        </div>
      </div>
      <div
        class="gist-meta"
      >
        <p>
          Created at: Invalid Date
        </p>
        <p
          style="margin-left: 20px; color: rgba(0, 0, 0, 0.3);"
        >
          |
        </p>
        <p
          style="margin-left: 20px;"
        >
          Last updated: Invalid Date
        </p>
      </div>
      <div
        class="gist-heading"
      >
        <p>
          Generated by: user1
        </p>
      </div>
      <div
        class="gist-files"
      >
        <p>
          Files:
        </p>
        <ul />
      </div>
    </div>
    <div
      class="gist"
    >
      <div
        class="gist-header"
      >
        <div
          class="user-info"
        >
          <img
            alt="user2"
            class="user-avatar"
          />
          <h2>
            user2
          </h2>
        </div>
        <div
          class="user-details"
        >
          <div
            class="gist-links"
          >
            <a
              rel="noreferrer"
              target="_blank"
            >
              <span
                class="octicon octicon-file"
              />
                Files
            </a>
            <a
              rel="noreferrer"
              target="_blank"
            >
              <span
                class="octicon octicon-repo-forked"
              />
                Forks
            </a>
            <a
              rel="noreferrer"
              target="_blank"
            >
              <span
                class="octicon octicon-comment"
              />
                Comments
            </a>
            <a
              rel="noreferrer"
              target="_blank"
            >
              <span
                class="octicon octicon-star"
              />
                Stars
            </a>
          </div>
        </div>
      </div>
      <div
        class="gist-meta"
      >
        <p>
          Created at: Invalid Date
        </p>
        <p
          style="margin-left: 20px; color: rgba(0, 0, 0, 0.3);"
        >
          |
        </p>
        <p
          style="margin-left: 20px;"
        >
          Last updated: Invalid Date
        </p>
      </div>
      <div
        class="gist-heading"
      >
        <p>
          Generated by: user2
        </p>
      </div>
      <div
        class="gist-files"
      >
        <p>
          Files:
        </p>
        <ul />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
